File: ict/feature.rb
#!/usr/bin/env ruby

def output_data(seq, fo, templates)
    seq.each{|h|
        # Extract more characteristics of the input sequence
        head = h['w'][0].clone
        h['iu'] = (h['w'] and head.downcase! != nil).to_s
    }

    for t in 0...seq.length do
        fo.print(seq[t]['y'])
        templates.each{ |template|
            atr = apply_template(seq, t, template)
            if atr != nil
                fo.print("\t"+escape(atr))
            end
        }
        fo.print("\n")
    end
    fo.print("\n")
end

def apply_template(seq, c, template)
    # fo[0]=field, fo[1]=offset
    name = template.map{|fo| 
        if fo != nil
            "#{fo[0]}[#{fo[1]}]"
        end
    }.compact.join("|")

    values = template.map{|fo|
        if fo != nil
            pt = c + fo[1]
            if pt.between?(0,seq.length-1)
                seq[pt][fo[0]]
            else
                false
            end
        end
    }

    if values.include?(false) 
        return nil
    else
        return name + "=" + values.compact.join("|")
    end

end

def escape(src)
    return src.gsub(/:/, "__COLON__")
end

if __FILE__ == $0
    fi = STDIN
    fo = STDOUT

    labels = ['w', 'pos', 'chk', 'iu']

    templates = []
    for l in labels do
        (-2...3).each{|i| templates.push([[l,i],nil]) }
        (-2...2).each{|i| templates.push([[l,i],[l,i+1]]) }
    end

    fnames = ['y', 'w', 'pos', 'chk']
    seq = []
    STDIN.each_line do |line|
        l = line.strip
        if l == ""
            output_data(seq, fo, templates)
            seq = []
        else
            fields = l.split("\t")
            if fields.length != fnames.length
                raise "format error: " + line
            else
                h = {}
                fnames.zip(fields).each{|p|
                    h[p[0]] = p[1]
                }
                seq.push(h)
            end
        end
    end
end

File: ict/feature.py
#!/usr/bin/env python

import sys

def readiter(fi, names=('y', 'w', 'pos', 'chk'), sep='\t'):
    seq = []
    for line in fi:
        line = line.strip('\n')
        if not line:
            yield seq
            seq = []
        else:
            fields = line.split(sep)
            if len(fields) != len(names):
                raise ValueError(
                    'Each line must have %d fields: %s\n' % (len(names), line))
            seq.append(dict(zip(names, tuple(fields))))

def apply_template(seq, t, template):
    name = '|'.join(['%s[%d]' % (f, o) for f, o in template])
    values = []
    for field, offset in template:
        p = t + offset
        if p not in range(len(seq)):
            return None
        values.append(seq[p][field])
    return '%s=%s' % (name, '|'.join(values))

def escape(src):
    return src.replace(':', '__COLON__')

if __name__ == '__main__':
    fi = sys.stdin
    fo = sys.stdout

    templates = []
    templates += [(('w', i),) for i in range(-2, 3)]
    templates += [(('w', i), ('w', i+1)) for i in range(-2, 2)]
    templates += [(('pos', i),) for i in range(-2, 3)]
    templates += [(('pos', i), ('pos', i+1)) for i in range(-2, 2)]
    templates += [(('chk', i),) for i in range(-2, 3)]
    templates += [(('chk', i), ('chk', i+1)) for i in range(-2, 2)]
    templates += [(('iu', i),) for i in range(-2, 3)]
    templates += [(('iu', i), ('iu', i+1)) for i in range(-2, 2)]

    for seq in readiter(fi):
        for v in seq:
            # Extract more characteristics of the input sequence
            v['iu'] = str(v['w'] and v['w'][0].isupper())

        for t in range(len(seq)):
            fo.write(seq[t]['y'])
            for template in templates:
                attr = apply_template(seq, t, template)
                if attr is not None:
                    fo.write('\t%s' % escape(attr))
            fo.write('\n')
        fo.write('\n')
